i18n/management/commands/publish_i18n.py

```diff
@@ -1,56 +1,50 @@
-import json
-import os
-
+# pylint: disable=missing-docstring,too-few-public-methods,no-self-use
 import django.apps
+
 from django.conf import settings
-from django.core.management.base import BaseCommand, CommandError
+from django.core.management.base import BaseCommand
 from django.utils import translation
 
-from i18n.models import Internationalizable
-from i18n.utils import print_clear
-
-from django_slack import slack_message
-
-def should_publish_model(model):
-    is_internationalizable = issubclass(model, Internationalizable)
-    is_not_proxy = not model._meta.proxy
-    has_publish_operation = hasattr(model, 'publish') or hasattr(model, 'publish_pdfs')
-    return is_internationalizable and is_not_proxy and has_publish_operation
+from i18n.management.utils import log, should_sync_model
 
-def log(message):
-    print(message)
-    slack_message('slack/message.slack', {
-        'message': message
-    })
 
 class Command(BaseCommand):
+    def should_publish_model(self, model):
+        model_has_publish_operation = hasattr(model, 'publish') or hasattr(model, 'publish_pdfs')
+        return should_sync_model(model) and model_has_publish_operation
+
     def handle(self, *args, **options):
-        log("Publishing translated content")
-        models = [model for model in django.apps.apps.get_models() if should_publish_model(model)]
+        log("I18n Sync Step 4 of 4: Publish translated content to S3")
+
+        models = [
+            model for model in django.apps.apps.get_models()
+            if self.should_publish_model(model)
+        ]
         log("Models to publish: %s" % ', '.join(model.__name__ for model in models))
 
+        language_codes = [
+            language_code for language_code, _ in settings.LANGUAGES
+            if language_code != settings.LANGUAGE_CODE
+        ]
+        log("Languages to publish: %s" % ', '.join(language_codes))
+
         for model_index, model in enumerate(models):
             name = model.__name__
             log("Publishing %s (%s/%s)" % (name, model_index + 1, len(models)))
             objects = model.get_i18n_objects()
             total = objects.count()
             success_count = 0
 
-            for index, obj in enumerate(objects.all()):
+            for obj in objects.all():
                 if not obj.should_be_translated:
                     continue
 
-                original_lang = translation.get_language()
-                for language_code, _ in settings.LANGUAGES:
-                    if language_code == settings.LANGUAGE_CODE:
-                        continue
-
+                for language_code in language_codes:
                     translation.activate(language_code)
                     if hasattr(obj, 'publish'):
                         list(obj.publish(silent=True))
                     if hasattr(obj, 'publish_pdfs'):
                         list(obj.publish_pdfs(silent=True))
```

Comment:
Is it worth logging failures as well?

Reply:
right now, failures in the publish operation are already noisy! A lot of the motivation here was to clear up the _other_ sources of noise, so those errors stand out more.

I would _also_ like to add the ability to track specifically whether or not the operations we're kicking off at various parts of this process (the up/downloads from crowdin, the redaction/restoration, the publish, etc) succeeded or not, and log those explicitly. That's part of what I'm going to be working on later this week.

Reply:
Sounds good. Thanks for the explanation!